src/main/scala/org/economicsl/auctions/orderbooks/FourHeapOrderBook.scala

```diff
@@ -0,0 +1,88 @@
+/*
+Copyright 2017 EconomicSL
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package org.economicsl.auctions.orderbooks
+
+import org.economicsl.auctions._
+
+
+class FourHeapOrderBook[T <: Tradable] private(matchedOrders: MatchedOrders[T], unMatchedOrders: UnMatchedOrders[T]) {
+
+  def - (order: LimitAskOrder[T]): FourHeapOrderBook[T] = {
```

Comment:
@bherd-rb Yes it is possible to pass in a sub-type of `LimitAskOrder` but without type parameter there is no way to make the behavior of `FourHeapOrderBook` depend on the specific sub-type of `LimitAskOrder`.  In this case I think we are OK, because the `FourHeapOrderBook` algorithm is bespoke for `SinglePricePoint` orders.

I have other, more generic implementation of `OrderBook` in for use cases where this restriction might not be appropriate.